File: app/controllers/ReportController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.ExpressionList;
import io.ebean.text.PathProperties;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Base64;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import javax.inject.Inject;
import models.Course;
import models.Exam;
import models.ExamEnrolment;
import models.ExamRoom;
import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;
import play.libs.Json;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.ExamRecordSanitizer;
import util.excel.ExcelBuilder;

public class ReportController extends BaseController {

    private static final String XLSX_MIME = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    private final ExcelBuilder excelBuilder;

    @Inject
    public ReportController(ExcelBuilder excelBuilder) {
        this.excelBuilder = excelBuilder;
    }

    @Restrict({ @Group("ADMIN") })
    public Result listDepartments() {
        List<Course> courses = DB.find(Course.class).where().isNotNull("department").findList();
        Set<String> departments = courses.stream().map(Course::getDepartment).collect(Collectors.toSet());
        ObjectNode node = Json.newObject();
        ArrayNode arrayNode = node.putArray("departments");
        departments.forEach(arrayNode::add);
        return ok(Json.toJson(node));
    }

    private <T> ExpressionList<T> applyFilters(
        ExpressionList<T> query,
        String deptFieldPrefix,
        String dateField,
        String dept,
        String start,
        String end
    ) {
        ExpressionList<T> result = query;
        if (dept != null) {
            List<String> depts = Arrays.asList(dept.split(","));
            result = result.in(String.format("%s.department", deptFieldPrefix), depts);
        }
        if (start != null) {
            DateTime startDate = DateTime.parse(start, ISODateTimeFormat.dateTimeParser());
            result = result.ge(dateField, startDate.toDate());
        }
        if (end != null) {
            DateTime endDate = DateTime.parse(end, ISODateTimeFormat.dateTimeParser()).plusDays(1);
            result = result.lt(dateField, endDate.toDate());
        }
        return result;
    }

    @Restrict({ @Group("ADMIN") })
    public Result getExamParticipations(Optional<String> dept, Optional<String> start, Optional<String> end) {
        PathProperties pp = PathProperties.parse(
            "noShow, exam(created, course(department)), " +
            "externalExam(started), reservation(machine(room(id, name, outOfService)))"
        );
        ExpressionList<ExamEnrolment> el = DB.find(ExamEnrolment.class).where();
        el.apply(pp);
        List<ExamEnrolment> enrolments = el
            .or()
            .ne("exam.state", Exam.State.PUBLISHED)
            .isNotNull("externalExam.started")
            .endOr()
            .isNotNull("reservation.machine")
            .ne("noShow", true)
            .findList()
            .stream()
            .filter(ee -> applyEnrolmentFilter(ee, dept, start, end))
            .toList();
        Map<String, List<Participation>> roomMap = new HashMap<>();
        for (ExamEnrolment enrolment : enrolments) {
            ExamRoom room = enrolment.getReservation().getMachine().getRoom();
            String key = String.format("%d___%s", room.getId(), room.getName());
            if (!roomMap.containsKey(key)) {
                roomMap.put(key, new ArrayList<>());
            }
            DateTime examStart = enrolment.getExternalExam() != null
                ? enrolment.getExternalExam().getStarted()
                : enrolment.getExam().getCreated();
            roomMap.get(key).add(new Participation(examStart));
        }
        // Fill in the rooms that have no associated participations
        List<ExamRoom> rooms = DB.find(ExamRoom.class).where().eq("outOfService", false).findList();
        for (ExamRoom room : rooms) {
            String key = String.format("%d___%s", room.getId(), room.getName());
            if (!roomMap.containsKey(key)) {
                // empty list indicates no participations
                roomMap.put(key, new ArrayList<>());
            }
        }
        return ok(Json.toJson(roomMap));
    }

    // DTO for minimizing output from this API
    private static class Participation {

        String date;

        Participation(DateTime date) {
            this.date = ISODateTimeFormat.dateTime().print(date);
        }

        public String getDate() {
            return date;
        }
    }

    private static class ExamInfo {

        String name;
        Integer participations;
        String state;

        public String getName() {
            return name;
        }

        public Integer getParticipations() {
            return participations;
        }

        public String getState() {
            return state;
        }
    }

    private boolean applyExamFilter(Exam e, Optional<String> start, Optional<String> end) {
        boolean result = e.getState().ordinal() > Exam.State.PUBLISHED.ordinal() && e.getExamParticipation() != null;
        DateTime created = e.getCreated();
        if (start.isPresent()) {
            DateTime startDate = DateTime.parse(start.get(), ISODateTimeFormat.dateTimeParser());
            result = result && startDate.isBefore(created);
        }
        if (end.isPresent()) {
            DateTime endDate = DateTime.parse(end.get(), ISODateTimeFormat.dateTimeParser()).plusDays(1);
            result = result && endDate.isAfter(created);
        }
        return result;
    }

    private boolean applyEnrolmentFilter(
        ExamEnrolment ee,
        Optional<String> dept,
        Optional<String> start,
        Optional<String> end
    ) {
        DateTime created = ee.getExam() != null ? ee.getExam().getCreated() : ee.getExternalExam().getStarted();
        boolean result = true;
        if (dept.isPresent()) {
            if (ee.getExternalExam() != null) {
                return false;
            }
            List<String> depts = Arrays.asList(dept.get().split(","));
            Course course = ee.getExam().getCourse();
            result = course != null && depts.contains(course.getDepartment());
        }
        if (start.isPresent()) {
            DateTime startDate = DateTime.parse(start.get(), ISODateTimeFormat.dateTimeParser());
            result = result && startDate.isBefore(created);
        }
        if (end.isPresent()) {
            DateTime endDate = DateTime.parse(end.get(), ISODateTimeFormat.dateTimeParser()).plusDays(1);
            result = result && endDate.isAfter(created);
        }
        return result;
    }

    @Restrict({ @Group("ADMIN") })
    public Result getPublishedExams(Optional<String> dept, Optional<String> start, Optional<String> end) {
        ExpressionList<Exam> query = DB
            .find(Exam.class)
            .fetch("course", "code")
            .where()
            .isNull("parent")
            .isNotNull("course")
            .disjunction()
            .eq("state", Exam.State.PUBLISHED)
            .eq("state", Exam.State.DELETED)
            .eq("state", Exam.State.ARCHIVED)
            .endJunction();
        query = applyFilters(query, "course", "created", dept.orElse(null), null, null);
        Set<Exam> exams = query.findSet();
        List<ExamInfo> infos = new ArrayList<>();
        for (Exam exam : exams) {
            ExamInfo info = new ExamInfo();
            info.name = String.format("[%s] %s", exam.getCourse().getCode(), exam.getName());
            info.participations = (int) exam.getChildren().stream().filter(e -> applyExamFilter(e, start, end)).count();
            infos.add(info);
        }
        return ok(Json.toJson(infos));
    }

    @Restrict({ @Group("ADMIN") })
    public Result getReservations(Optional<String> dept, Optional<String> start, Optional<String> end) {
        ExpressionList<ExamEnrolment> query = DB.find(ExamEnrolment.class).where();
        query =
            applyFilters(
                query,
                "enrolment.exam.course",
                "startAt",
                dept.orElse(null),
                start.orElse(null),
                end.orElse(null)
            );
        Set<ExamEnrolment> enrolments = query.findSet();
        long noShows = enrolments.stream().filter(ExamEnrolment::isNoShow).count();
        long appearances = enrolments.size() - noShows;
        return ok(Json.newObject().put("noShows", noShows).put("appearances", appearances));
    }

    @Restrict({ @Group("ADMIN") })
    public Result getResponses(Optional<String> dept, Optional<String> start, Optional<String> end) {
        ExpressionList<Exam> query = DB.find(Exam.class).where().isNotNull("parent").isNotNull("course");
        query = applyFilters(query, "course", "created", dept.orElse(null), start.orElse(null), end.orElse(null));
        Set<Exam> exams = query.findSet();
        long aborted = exams.stream().filter(e -> e.getState() == Exam.State.ABORTED).count();
        long assessed = exams
            .stream()
            .filter(e ->
                e.hasState(
                    Exam.State.GRADED,
                    Exam.State.GRADED_LOGGED,
                    Exam.State.ARCHIVED,
                    Exam.State.REJECTED,
                    Exam.State.DELETED
                )
            )
            .count();
        long unAssessed = exams
            .stream()
            .filter(e ->
                e.hasState(
                    Exam.State.INITIALIZED,
                    Exam.State.STUDENT_STARTED,
                    Exam.State.REVIEW,
                    Exam.State.REVIEW_STARTED
                )
            )
            .count();
        JsonNode node = Json
            .newObject()
            .put("aborted", aborted)
            .put("assessed", assessed)
            .put("unAssessed", unAssessed);
        return ok(node);
    }

    @With(ExamRecordSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result exportExamQuestionScoresAsExcel(Long examId, Http.Request request) {
        Collection<Long> childIds = request.attrs().get(Attrs.ID_COLLECTION);
        ByteArrayOutputStream bos;
        try {
            bos = excelBuilder.buildScoreExcel(examId, childIds);
        } catch (IOException | RuntimeException e) {
            return internalServerError("i18n_error_creating_csv_file");
        }
        return ok(Base64.getEncoder().encodeToString(bos.toByteArray()))
            .withHeader("Content-Disposition", "attachment; filename=\"exam_records.xlsx\"")
            .as(XLSX_MIME);
    }
}

File: app/controllers/admin/ReportController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.admin

import io.ebean.text.PathProperties
import io.ebean.{DB, ExpressionList}
import miscellaneous.excel.ExcelBuilder
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.enrolment.{ExamEnrolment, Reservation}
import models.exam.{Course, Exam}
import models.facility.ExamRoom
import models.user.Role
import org.joda.time.DateTime
import org.joda.time.format.ISODateTimeFormat
import play.api.libs.json.{Json, Writes}
import play.api.mvc.*
import security.scala.Auth.authorized
import security.scala.AuthExecutionContext
import system.AuditedAction
import validation.scala.core.{ScalaAttrs, Validators}
import validation.scala.CommaJoinedListValidator

import java.util.Base64
import javax.inject.Inject
import scala.jdk.CollectionConverters.*
import scala.util.{Failure, Success, Try}

class ReportController @Inject() (
    val controllerComponents: ControllerComponents,
    val excelBuilder: ExcelBuilder,
    validators: Validators,
    audited: AuditedAction,
    implicit val ec: AuthExecutionContext
) extends BaseController
    with DbApiHelper
    with JavaApiHelper:

  case class ExamInfo(name: String, participations: Int)
  object ExamInfo:
    implicit val writes: Writes[ExamInfo] = Json.writes[ExamInfo]

  case class Participation(date: String)
  object Participation:
    def apply(dateTime: DateTime): Participation =
      Participation(ISODateTimeFormat.dateTime.print(dateTime))
    implicit val writes: Writes[Participation] = Json.writes[Participation]

  private val XLSX_MIME = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"

  def listDepartments: Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val departments = DB
        .find(classOf[Course])
        .where()
        .isNotNull("department")
        .distinct
        .map(_.getDepartment)
      Ok(Json.obj("departments" -> Json.toJson(departments)))
    }

  def listParticipations(dept: Option[String], start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val pp = PathProperties.parse(
        """noShow,
          |exam(created,
          |  course(department)
          |),
          |externalExam(started),
          |reservation(
          |  machine(
          |    room(id, name, outOfService)
          |  )
          |)
          |)""".stripMargin
      )
      val enrolments = DB
        .find(classOf[ExamEnrolment])
        .where
        .apply(pp)
        .where()
        .or()
        .ne("exam.state", Exam.State.PUBLISHED)
        .isNotNull("externalExam.started")
        .endOr()
        .isNotNull("reservation.machine")
        .ne("noShow", true)
        .distinct
      val roomMap = enrolments
        .groupBy { enrolment =>
          val room = enrolment.getReservation.getMachine.getRoom
          s"${room.getId}___${room.getName}"
        }
        .view
        .mapValues { enrolmentList =>
          enrolmentList.map { enrolment =>
            val examStart = Option(enrolment.getExternalExam)
              .map(_.getStarted)
              .getOrElse(enrolment.getExam.getCreated)
            Participation(examStart)
          }
        }
        .toMap

      // Fill in rooms with no participations

      val allRooms = DB.find(classOf[ExamRoom]).where().eq("outOfService", false).list
      val completeRoomMap = allRooms.foldLeft(roomMap) { (map, room) =>
        val key = s"${room.getId}___${room.getName}"
        if (map.contains(key)) map else map + (key -> Set.empty[Participation])
      }

      Ok(Json.toJson(completeRoomMap))
    }

  def listReservations(dept: Option[String], start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val query = DB.find(classOf[ExamEnrolment]).where()
      val enrolments =
        withFilters(query, "exam.course", "reservation.startAt", dept, start, end).distinct
      val (noShows, appearances) = enrolments.partition(_.isNoShow) match
        case (a, b) => (a.size, b.size)
      Ok(Json.obj("noShows" -> noShows, "appearances" -> appearances))
    }

  def listIopReservations(dept: Option[String], start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val pp    = PathProperties.parse("*, enrolment(noShow, externalExam(finished)), externalReservation(*)")
      val query = DB.find(classOf[Reservation]).where().apply(pp)
      val el    = query.where().isNotNull("externalRef")
      val reservations = withFilters(el, "enrolment.exam.course", "startAt", dept, start, end).distinct.filter(r =>
        Option(r.getExternalOrgName).isDefined || Option(r.getExternalReservation).isDefined
      )
      Ok(reservations.asJson)
    }

  def listPublishedExams(dept: Option[String], start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val query = DB
        .find(classOf[Exam])
        .fetch("course", "code")
        .where()
        .isNull("parent")
        .isNotNull("course")
        .in("state", Exam.State.PUBLISHED, Exam.State.DELETED, Exam.State.ARCHIVED)
      val exams = withFilters(query, "course", "created", dept, start, end).distinct

      def examFilter(exam: Exam) =
        val created          = exam.getCreated
        val hasValidState    = exam.getState.ordinal() > Exam.State.PUBLISHED.ordinal()
        val hasParticipation = Option(exam.getExamParticipation).isDefined
        hasValidState &&
        hasParticipation &&
        start.forall(s => DateTime.parse(s, ISODateTimeFormat.dateTimeParser()).isBefore(created)) &&
        end.forall(e => DateTime.parse(e, ISODateTimeFormat.dateTimeParser()).plusDays(1).isAfter(created))

      val infos =
        exams.map(e => ExamInfo(s"[${e.getCourse.getCode}] ${e.getName}", e.getChildren.asScala.count(examFilter)))
      Ok(Json.toJson(infos))
    }

  def listResponses(dept: Option[String], start: Option[String], end: Option[String]): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val query   = DB.find(classOf[Exam]).where().isNotNull("parent").isNotNull("course")
      val exams   = withFilters(query, "course", "created", dept, start, end).distinct
      val aborted = exams.count(e => e.getState == Exam.State.ABORTED)
      val assessed = exams.count(e =>
        e.hasState(
          Exam.State.GRADED,
          Exam.State.GRADED_LOGGED,
          Exam.State.ARCHIVED,
          Exam.State.REJECTED,
          Exam.State.DELETED
        )
      )
      val unassessed = exams.count(e =>
        e.hasState(
          Exam.State.INITIALIZED,
          Exam.State.STUDENT_STARTED,
          Exam.State.REVIEW,
          Exam.State.REVIEW_STARTED
        )
      )
      Ok(Json.obj("aborted" -> aborted, "assessed" -> assessed, "unassessed" -> unassessed))
    }

  def exportExamQuestionScoresAsExcel(examId: Long): Action[AnyContent] =
    Action
      .andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.TEACHER)))
      .andThen(validators.validated(CommaJoinedListValidator))
      .andThen(audited) { request =>
        val childIds = request.attrs(ScalaAttrs.ID_LIST)
        Try(excelBuilder.buildScoreExcel(examId, childIds)) match
          case Success(bos) =>
            val encoded = Base64.getEncoder.encodeToString(bos.toByteArray)
            Ok(encoded)
              .withHeaders("Content-Disposition" -> "attachment; filename=\"exam_records.xlsx\"")
              .as(XLSX_MIME)
          case Failure(_) =>
            InternalServerError("i18n_error_creating_csv_file")
      }

  private def withFilters[T](
      query: ExpressionList[T],
      deptFieldPrefix: String,
      dateField: String,
      depts: Option[String],
      start: Option[String],
      end: Option[String]
  ): ExpressionList[T] =
    val withDept = depts.fold(query) { d =>
      val deptList = d.split(",").toList
      query.in(s"$deptFieldPrefix.department", deptList*)
    }
    val withStart = start.fold(withDept) { s =>
      val startDate = DateTime.parse(s, ISODateTimeFormat.dateTimeParser())
      withDept.ge(dateField, startDate.toDate)
    }
    end.fold(withStart) { e =>
      val endDate = DateTime.parse(e, ISODateTimeFormat.dateTimeParser()).plusDays(1)
      withStart.lt(dateField, endDate.toDate)
    }
